cumulus_library/statistics/psm.py

```diff
@@ -0,0 +1,247 @@
+        )
+        neg_query = get_distinct_ids(
+            source_refs,
+            self.config.neg_source_table,
+            join_id=self.config.primary_ref,
+            filter_table=self.config.pos_source_table,
+        )
+        neg = self._get_sampled_ids(
+            cursor,
+            schema,
+            neg_query,
+            self.config.neg_sample_size,
+            self.config.dependent_variable,
+            0,
+        )
+        cohort = pandas.concat([pos, neg])
+
+        # Replace table (if it exists)
+        # TODO - replace with timestamp prepended table
+        drop = get_drop_view_table(
+            f"{self.config.pos_source_table}_sampled_ids", "TABLE"
+        )
+        cursor.execute(drop)
+        ctas_query = get_ctas_query_from_df(
+            schema,
+            f"{self.config.pos_source_table}_sampled_ids",
+            cohort,
+        )
+        self.queries.append(ctas_query)
+        # TODO - replace with timestamp prepended table
+        drop = get_drop_view_table(self.config.target_table, "TABLE")
+        cursor.execute(drop)
+        dataset_query = get_create_covariate_table(
+            target_table=self.config.target_table,
+            pos_source_table=self.config.pos_source_table,
+            neg_source_table=self.config.neg_source_table,
+            primary_ref=self.config.primary_ref,
+            dependent_variable=self.config.dependent_variable,
+            join_cols_by_table=self.config.join_cols_by_table,
+            count_ref=self.config.count_ref,
+            count_table=self.config.count_table,
+        )
+        self.queries.append(dataset_query)
+
+    def generate_psm_analysis(self, cursor: object, schema: str):
+        """Runs PSM statistics on generated tables"""
+        df = cursor.execute(f"select * from {self.config.target_table}").as_pandas()
+        symptoms_dict = self._get_symptoms_dict(self.config.classification_json)
+        for dependent_variable, codes in symptoms_dict.items():
+            df[dependent_variable] = df["code"].apply(lambda x: 1 if x in codes else 0)
+        df = df.drop(columns="code")
+        # instance_count present but unused for PSM if table contains a count_ref input
+        # (it's intended for manual review)
+        df = df.drop(columns="instance_count", errors="ignore")
+
+        columns = []
+        if self.config.join_cols_by_table is not None:
+            for table_key in self.config.join_cols_by_table:
+                for column in self.config.join_cols_by_table[table_key][
+                    "included_cols"
+                ]:
+                    if len(column) == 2:
+                        columns.append(column[1])
+                    else:
+                        columns.append(column[0])
+
+        for column in columns:
+            encoded_df = pandas.get_dummies(df[column])
+            df = pandas.concat([df, encoded_df], axis=1)
+            df = df.drop(column, axis=1)
+        df = df.reset_index()
```

Comment:
A block like this could use a comment about why it's doing what it's doing. My rough attempt: you're replacing each column with a dummy version of that column. But even after reading the pandas docs on `get_dummies`, I'm not :100: on what that means.

Reply:
yeah i'll add something - this is converting to a 1-hot encoding for all values of that column, basically pivoting the column values to new column headers.

Reply:
updated with some explanitory text